ArjunaCore/arjuna/classes/com/arjuna/ats/internal/arjuna/objectstore/slot/SlotStoreAdaptor.java

```diff
@@ -0,0 +1,349 @@
+        store = new SlotStore(envBean);
+    }
+
+    // used for beans wiring type bootstrap when running embedded.
+    public SlotStoreAdaptor(SlotStore slotStore) {
+        this.store = slotStore;
+    }
+
+    /**
+     * Obtain all of the Uids for a specified type.
+     *
+     * @param typeName       The type to scan for.
+     * @param foundInstances The object state in which to store the Uids
+     * @param matchState     The file type to look for (e.g., committed, shadowed). [StateStatus]
+     * @return <code>true</code> if no errors occurred, <code>false</code>
+     * otherwise.
+     */
+    @Override
+    public boolean allObjUids(String typeName, InputObjectState foundInstances, int matchState) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allObjUids(" + typeName + ", " + matchState + ")");
+        }
+
+        SlotStoreKey key = new SlotStoreKey(null, typeName, matchState);
+
+        SlotStoreKey[] matchingKeys = store.getMatchingKeys(key);
+
+        OutputObjectState buffer = new OutputObjectState();
+
+        try {
+            if (matchingKeys != null) {
+                for (SlotStoreKey matchingKey : matchingKeys) {
+                    if (matchingKey != null) {
+                        UidHelper.packInto(matchingKey.getUid(), buffer);
+                    }
+                }
+            }
+            UidHelper.packInto(Uid.nullUid(), buffer);
+        } catch (IOException e) {
+            throw new ObjectStoreException(e);
+        }
+
+        foundInstances.setBuffer(buffer.buffer());
+
+        return true;
+    }
+
+    @Override
+    public boolean allObjUids(String typeName, InputObjectState foundInstances) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allObjUids(" + typeName + ")");
+        }
+
+        return allObjUids(typeName, foundInstances, StateStatus.OS_UNKNOWN);
+    }
+
+    /**
+     * Obtain all types of objects stored in the object store.
+     *
+     * @param foundTypes The state in which to store the types.
+     * @return <code>true</code> if no errors occurred, <code>false</code>
+     * otherwise.
+     */
+    @Override
+    public boolean allTypes(InputObjectState foundTypes) throws ObjectStoreException {
+        if (tsLogger.logger.isTraceEnabled()) {
+            tsLogger.logger.trace("SlotStore.allTypes()");
+        }
+
+        String[] knownTypes = store.getKnownTypes(); // may contain trailing null elements
+        Set<String> typeSet = new HashSet<>();
+
+        if (knownTypes == null || knownTypes.length == 0) {
+            return true;
+        }
+
+        OutputObjectState buffer = new OutputObjectState();
+
+        try {
+            for (String typeName : knownTypes) {
+                if (typeName == null) {
+                    continue;
+                }
+
+                if (typeName.startsWith("/")) {
```

Comment:
Remark: ArjunaCore does not document that typename hierarchies are delimited by slashes
Remark: could you not leverage java.nio.files.Path to split the type name for you instead of using the following tricky code

Reply:
ArjunaCore uses leading delimiters in some calls and not in others, considering the typename to be equivalent with or without, which works ok when backed by the file system as the file path code fixes them up when concatenating, but not when backed by non-fs things like the string indexing here. Basicall "/foo".equals("foo") is assumed by the core to be true even though it's not, because in practice new File(baseDIr, "/foo").equals( new File(baseDIr, "foo") ) is true.
Took an age to track that down when building the Hornetq version. It's another thing that got copied in and could be tweaked, but probably isn't worth the effort. More complex parsing is overkill as we don't care about / within the typename, only at the start.

Reply:
I was asking more about using java.nio.files.Path to split the type name instead of your hand rolled complex parsing. But if you prefer to keep the complex version then fine and we will maintain it for you ;)

Reply:
Actually using java.nio.* is more complex and slower. Please ignore my comment.